### PR TITLE
ROS-243 Create primitives for math functions

### PR DESCRIPTION
**Pull request by AbnerZheng**
@tymm  Please have a look at this PR for tickets ROS-243.

### CODE REVIEW

File: roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala [hunk 1]
```diff
@@ -318,4 +322,245 @@ class PrimSpec extends FlatSpec with Matchers {
...
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(Math.E)))
+    flLog.fn(newCtxt) should be(Right(RblFloat(1)))
+  }
+
+  it should "fail for non-RblFloat arguments" in {
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(Ob.NIV))
+    flLog.fn(newCtxt) should be('left)
+  }
+
+  "flLog10" should "correctly return the base 10 logarithm of a double value." in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala]: `"flLog" should "return result of common logarithm for input value"`.

File: roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala [hunk 2]
```diff
@@ -318,4 +322,245 @@ class PrimSpec extends FlatSpec with Matchers {
...
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(100.0)))
+    flLog10.fn(newCtxt) should be(Right(RblFloat(2.0)))
+  }
+
+  it should "fail for  non-RblFloat arguments" in {
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(Ob.NIV))
+    flLog10.fn(newCtxt) should be('left)
+  }
+
+  "flCeil" should "correctly return ceil(input_value)" in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala]: `return ceiling of input value`

File: roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala [hunk 3]
```diff
@@ -318,4 +322,245 @@ class PrimSpec extends FlatSpec with Matchers {
...
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(2.1)))
+    flCeil.fn(newCtxt) should be(Right(RblFloat(3.0)))
+  }
+
+  it should "fail for  non-RblFloat arguments" in {
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(Ob.NIV))
+    flCeil.fn(newCtxt) should be('left)
+  }
+
+  "flFloor" should "correctly return floor(input_value)" in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala]: `return floor of input value`

File: roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala [hunk 4]
```diff
@@ -318,4 +322,245 @@ class PrimSpec extends FlatSpec with Matchers {
...
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(2.1)))
+    flFloor.fn(newCtxt) should be(Right(RblFloat(2.0)))
+  }
+
+  it should "fail for  non-RblFloat arguments" in {
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(Ob.NIV))
+    flFloor.fn(newCtxt) should be('left)
+  }
+
+  "flAtan" should "correctly return the arc tangent of a value; the returned angle is in the range -pi/2 through pi/2. " in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala]: `return the arc tangent of the input value` and put the explanation about the range into the test case as a comment.

File: roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala [hunk 5]
```diff
@@ -318,4 +322,245 @@ class PrimSpec extends FlatSpec with Matchers {
...
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(Math.PI)))
+    flSin.fn(newCtxt) should be(Right(RblFloat(Math.sin(Math.PI))))
+  }
+
+  it should "fail for non-RblFloat arguments" in {
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(Ob.NIV))
+    flSin.fn(newCtxt) should be('left)
+  }
+
+  "flCos" should "correctly return Returns the trigonometric cosine of an angle." in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala]: `return Returns`?

File: roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala [hunk 6]
```diff
@@ -318,4 +322,245 @@ class PrimSpec extends FlatSpec with Matchers {
...
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(Math.PI)))
+    flCos.fn(newCtxt) should be(Right(RblFloat(Math.cos(Math.PI))))
+  }
+
+  it should "fail for non-RblFloat arguments" in {
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(Ob.NIV))
+    flFloor.fn(newCtxt) should be('left)
+  }
+
+  "flToFx" should "correctly convert the float value to integer" in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/PrimSpec.scala]: `convert RblFloat to Fixnum`.

File: roscala/src/main/scala/coop/rchain/rosette/prim/Number.scala [hunk 7]
```diff
@@ -479,16 +487,420 @@ object Number {
...
+      try {
+        Right(m / n)
+      } catch {
+        case _: ArithmeticException =>
+          Left(ArithmeticError)
+      }
+    }
+  }
+
+  object flLt extends Prim {
```
[the remark this answers is not in the document]
Comment (AbnerZheng) [reply roscala/src/main/scala/coop/rchain/rosette/prim/Number.scala]: Yes. As show in rossette, Number.cc#407, RblFloat has inconsistent behavior with Fixnum in comparing.

File: roscala/src/main/scala/coop/rchain/rosette/prim/Number.scala [hunk 8]
```diff
@@ -479,16 +487,420 @@ object Number {
...
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[RblFloat]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, RblBool] = {
+      val m = ctxt.argvec.elem.head.asInstanceOf[RblFloat]
+      val n = ctxt.argvec.elem(1).asInstanceOf[RblFloat]
+
+      try {
```
[the remark this answers is not in the document]
Comment (AbnerZheng) [reply roscala/src/main/scala/coop/rchain/rosette/prim/Number.scala]: Yes. Actually almost function in scala.math will not throw exception. I will clean these unneeded code.

[the remark this answers is not in the document]
Comment (AbnerZheng) [reply roscala/src/main/scala/coop/rchain/rosette/prim/Number.scala]: And I'll create another PR to clean unneeded `try catch` in the method of Fixnum.

File: roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(-2.1)))
+    flAbs.fn(newCtxt) should be(Right(RblFloat(2.1)))
+  }
+
+  it should "fail for non-RblFloat arguments" in {
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(1, Ob.NIV))
+    flAbs.fn(newCtxt) should be('left)
+  }
+
+  "flExp" should "Returns Euler's number e raised to the power of input value." in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala]: `return e to the power of the input value`.

File: roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(2.5)))
+    flExp.fn(newCtxt) should be(Right(RblFloat(math.exp(2.5))))
+  }
+
+  it should "fail for non-RblFloat arguments" in {
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(Ob.NIV))
+    flExp.fn(newCtxt) should be('left)
+  }
+
+  "flExpt" should "correctly return e to the power of the input value" in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala]: Seems like you mixed stuff up here.

File: roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(100.0)))
+    flLog10.fn(newCtxt) should be(Right(RblFloat(2.0)))
+  }
+
+  it should "fail for  non-RblFloat arguments" in {
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(Ob.NIV))
+    flLog10.fn(newCtxt) should be('left)
+  }
+
+  "flCeil" should "correctly return ceiling of input_value" in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala]: `input value`.

**Comment by AbnerZheng** [comment]
There is some related file being influenced when Fixnum and RblFloat move to separate class file.

**Comment by tymm** [comment]
Looks good but can you put the `Fixnum` and `RblFloat` case classes in separate files (`Fixnum.scala` and `RblFloat.scala` in the source code root directory)?

**Comment by tymm** [comment]
When I said that you should put the `Fixnum` and `RblFloat` case classes into separate files, I meant that you should keep the primitives for them in `prim/Fixnum.scala` and `prim/RblFloat.scala`. Sorry, should have explicitly said that.

File: roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala [hunk 12]
```diff
@@ -174,8 +174,8 @@ class RblFloatSpec extends FlatSpec with Matchers {
   }
 
   "flExpt" should "correctly return base-number raised to the power power-number" in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala]: `correctly return first argument to the power of second argument`.

File: roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala [hunk 13]
```diff
@@ -245,18 +245,18 @@ class RblFloatSpec extends FlatSpec with Matchers {
...
 
   it should "fail for non-RblFloat arguments" in {
     val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(Ob.NIV))
     flFloor.fn(newCtxt) should be('left)
   }
 
-  "flToFx" should "correctly convert the RblFloat value to Fixnum" in {
-    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(2.1)))
-    flToFx.fn(newCtxt) should be(Right(Fixnum(2)))
+  "flToFx" should "correctly convert the RblFloat value to RFixnum" in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala]: `Fixnum`.

File: roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala [hunk 14]
```diff
@@ -1,7 +1,7 @@
 package coop.rchain.rosette.prim
 
-import coop.rchain.rosette.RblFloat._
-import coop.rchain.rosette.{Ctxt, Fixnum, Ob, PC, RblBool, RblFloat, Tuple}
+import coop.rchain.rosette.prim.RblFloat._
+import coop.rchain.rosette.{Ctxt, Fixnum=>RFixnum, Ob, PC, RblBool, RblFloat=>RFloat, Tuple}
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala]: Please add whitespace before and after `=>`.

File: roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala [hunk 15]
```diff
@@ -214,8 +214,8 @@ class RblFloatSpec extends FlatSpec with Matchers {
   }
 
   "flFloor" should "correctly return floor of input value" in {
-    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(2.1)))
-    flFloor.fn(newCtxt) should be(Right(RblFloat(2.0)))
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RFloat(2.1)))
+    flFloor.fn(newCtxt) should be(Right(RFloat(2.0)))
   }
 
   it should "fail for  non-RblFloat arguments" in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala]: Remove whitespace.

File: roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala [hunk 16]
```diff
@@ -204,8 +204,8 @@ class RblFloatSpec extends FlatSpec with Matchers {
   }
 
   "flCeil" should "correctly return ceiling of input value" in {
-    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(2.1)))
-    flCeil.fn(newCtxt) should be(Right(RblFloat(3.0)))
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RFloat(2.1)))
+    flCeil.fn(newCtxt) should be(Right(RFloat(3.0)))
   }
 
   it should "fail for  non-RblFloat arguments" in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala]: Remove whitespace.

File: roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala [hunk 17]
```diff
@@ -224,9 +224,9 @@ class RblFloatSpec extends FlatSpec with Matchers {
...
 
   "flAtan" should "correctly return the arc tangent of a value " in {
-    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(1.0)))
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RFloat(1.0)))
     // the returned angle is in the range -pi/2 through pi/2.
-    flAtan.fn(newCtxt) should be(Right(RblFloat(Math.PI / 4)))
+    flAtan.fn(newCtxt) should be(Right(RFloat(Math.PI / 4)))
   }
 
   it should "fail for  non-RblFloat arguments" in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala]: Remove whitespace.

File: roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala [hunk 18]
```diff
@@ -194,8 +194,8 @@ class RblFloatSpec extends FlatSpec with Matchers {
   }
 
   "flLog10" should "correctly return result of common logarithm for input value" in {
-    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RblFloat(100.0)))
-    flLog10.fn(newCtxt) should be(Right(RblFloat(2.0)))
+    val newCtxt = ctxt.copy(nargs = 1, argvec = Tuple(RFloat(100.0)))
+    flLog10.fn(newCtxt) should be(Right(RFloat(2.0)))
   }
 
   it should "fail for  non-RblFloat arguments" in {
```
Comment (tymm) [inline roscala/src/test/scala/coop/rchain/rosette/prim/RblFloatSpec.scala]: Remove whitespace.

**Comment by pyrocto** [comment]
Code should be run through scalafmt; that should fix most if not all of your formatting problems.  Please also see https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards

**Pull request merged**